### ISSUE TITLE
[Bug] DeepSeek V4 fixed-seed decode intermittently diverges into unrelated JSON with and without MTP

### ISSUE THREAD

**Issue by vegetabledoww** [hw-native-sys/pypto-lib#1122]
## Summary

DeepSeek V4 Flash W8A8 intermittently produces different completions for identical requests even when the prompt, sampling parameters, and seed are fixed. Some divergent runs jump at output token 30 from a coherent Palace Museum description into unrelated shopping-product JSON.

This reproduces with both `--num-speculative-tokens 1` and `--num-speculative-tokens 0`, so MTP is not required to trigger it. The current evidence points to nondeterminism in the common decode/sampling path rather than an MTP-only correctness issue.

## Environment

- pypto-lib: `05dc1b96f6a57b053feb527effba31463f51c7eb`
- pypto-serving: `b4df24453ae108739cf6c4830ef692f129a366cf`
- pypto: `c2414da6b8f248c475cce8aee6a09bf568224f67`
- runtime: `15f5cbd922494c62444e75fa1585e39f86a64a78`
- PTO-ISA: `96ba706ce1697dd5febe107ee41a72b26e687b42`
- Model: DeepSeek V4 Flash W8A8 (`/data/models/dsv4-flash-w8a8` on the test host)
- Devices: `1,3,5,7,9,11,13,15`, DP=8, EP=8
- Task: `task_20260903_013905_1153719185`

## Reproduction

Start one service and reuse it for 16 sequential requests. Run the experiment once with K=1 and once with K=0:

```bash
python -m pypto_serving.cli \
  --model /data/models/dsv4-flash-w8a8 \
  --served-model-name deepseek-v4-flash \
  --backend npu --platform a2a3 \
  --devices 1,3,5,7,9,11,13,15 --dp 8 --ep 8 \
  --block-size 128 --max-model-len 260 \
  --max-num-seqs 8 --max-num-batched-tokens 1024 \
  --npu-memory-utilization 0.90 \
  --long-prefill-token-threshold 1024 \
  --num-speculative-tokens <0-or-1> \
  --no-enable-prefix-caching \
  --ring-dep-pool 16384 --ring-task-window 16384 \
  --ring-heap 1073741824
```

Send this identical `/v1/completions` request 16 times:

```json
{
  "model": "deepseek-v4-flash",
  "prompt": "紫禁城南北长961米，东西宽753米，四面围有高10米的城墙，城外有宽52米的护城河。紫禁城有四座城门，南面为午门，北面为神武门，东面为东华门，西面为西华门。",
  "max_tokens": 128,
  "temperature": 0.8,
  "top_p": 1.0,
  "top_k": 32,
  "seed": 42
}
```

The prompt is 64 tokens and every response contains 128 output tokens.

## Observed results

| Mode | Repeated-request result | Unrelated JSON divergence |
| --- | --- | --- |
| K=1 | 3 unique outputs; 14/16 exactly matched the dominant/reference output | 1/16 at output token 30; one additional divergence at token 104 |
| K=0 | 5 unique outputs; dominant output occurred 12/16 | 3/16 at output token 30 |

The K=0 `exact_match=0/16` field in the local artifact is not itself evidence of a failure: that comparison used the K=1 expected text, and the normal K=0 continuation differs from it at token 66. The relevant K=0 evidence is that identical K=0 requests produced five distinct outputs, including three unrelated JSON continuations.

All token-30 failures share this transition:

```text
...民间有九梁十八柱七十二条脊之说，形容其结构的复杂。", 
  "title": "故宫博物院 ..."
 },
 {
  "date": "...",
  "link": "https://www.smzdm.com/...",
  "description": "..."
```

The coherent/reference continuation at the same point is:

```text
...形容其结构的复杂。紫禁城内的建筑分为外朝和内廷两部分。外朝的中心为...
```

The JSON failure occurred on multiple physical devices in the K=0 run (devices 1, 11, and 13), so the observation does not isolate to one bad device. The server logs contain no runtime or hardware error.

For K=1, the JSON-divergent request had MTP acceptance `51/78 = 65.38%`. However, the separate token-104 divergence had acceptance `64/66 = 96.97%`, so low MTP acceptance is not a sufficient explanation.

## Compiler warning worth investigating

Both K=0 and K=1 server builds emit `ScalarWriteLineShared` for the shared sampling path:

```text
pl.write into 'mask_scores' from a runtime number of concurrent blocks
('sample_apply_top_k_spmd') in function 'apply_top_k' ... two blocks sharing
a line silently lose each other's stores.
```

The warning points to `models/deepseek_v4_flash_mtp/sample.py:458` and `:463` at this revision. This is an actionable lead because both modes use this top-k sampling path, but it is not yet proven to be the root cause. Similar warnings are also emitted for MoE dispatch buffers and CSA sparse-attention buffers, so those common decode boundaries remain candidates as well.

## Expected behavior

Repeated requests with the same model, prompt, sampling parameters, and seed should replay the same sampled token sequence. At minimum, they should not intermittently switch from a coherent completion into an unrelated JSON corpus fragment.

## Suggested next checks

1. Capture the main-model logits/top-k candidates around output token 30 for a normal and a JSON-divergent request.
2. Check whether serializing or cache-line-aligning the `mask_scores` writes removes the divergence.
3. If sampling inputs already differ, bisect the common decode path at HC pre/QKV, CSA, MoE, and LM head boundaries with fixed inputs.
4. Compare repeated runs using frozen decode inputs to separate kernel nondeterminism from request-state contamination.

## Related issues

- #951: previous DeepSeek V4 long-decode drift caused by RoPE profile and nondeterministic split-K `AtomicAdd`. This reproduces after that fix at pypto-lib `05dc1b9`.
- #1081: related INT8/top-k-boundary sensitivity in MTP CSA, but its frozen-input fixture is deterministic and the reproduction here also occurs with K=0.
- #929: previous persistent-worker MoE state contamination. Potentially related, but this case uses the same prompt repeatedly and reproduces across both K=0 and K=1.

Local artifacts on the test host:

- `/data/jinzongquan/pr187-ci-repro-b4df244/logs/long-completion-k1-odd-20260903-014353-results.json`
- `/data/jinzongquan/pr187-ci-repro-b4df244/logs/long-completion-k1-odd-20260903-014353-server.log`
- `/data/jinzongquan/pr187-ci-repro-b4df244/logs/long-completion-k0-odd-20260903-015403-results.json`
- `/data/jinzongquan/pr187-ci-repro-b4df244/logs/long-completion-k0-odd-20260903-015403-server.log`
